graylog2-server/src/main/java/org/graylog/integrations/inputs/paloalto11/PaloAlto11xCodec.java

```diff
@@ -0,0 +1,179 @@
+import org.graylog2.plugin.configuration.fields.DropdownField;
+import org.graylog2.plugin.inputs.annotations.ConfigClass;
+import org.graylog2.plugin.inputs.annotations.FactoryClass;
+import org.graylog2.plugin.inputs.codecs.Codec;
+import org.graylog2.plugin.inputs.codecs.CodecAggregator;
+import org.graylog2.plugin.journal.RawMessage;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+public class PaloAlto11xCodec implements Codec {
+    private static final Logger LOG = LoggerFactory.getLogger(PaloAlto11xCodec.class);
+    static final String CK_STORE_FULL_MESSAGE = "store_full_message";
+    static final String CK_TIMEZONE = "timezone";
+
+    public static final String NAME = "PaloAlto11x";
+    public static final String EVENT_SOURCE_PRODUCT_NAME = "PAN";
+    public static final String UNKNOWN = "unknown";
+
+    private final Configuration configuration;
+    private final MessageFactory messageFactory;
+    private final PaloAltoParser rawMessageParser;
+    private final DateTimeZone timezone;
+
+    @AssistedInject
+    public PaloAlto11xCodec(@Assisted Configuration configuration, PaloAltoParser rawMessageParser,
+                            MessageFactory messageFactory) {
+        this.configuration = configuration;
+        this.messageFactory = messageFactory;
+        String timezoneID = configuration.getString(CK_TIMEZONE);
+        // previously existing PA inputs after updating will not have a Time Zone configured, default to UTC
+        this.timezone = timezoneID != null ? DateTimeZone.forID(timezoneID) : DateTimeZone.UTC;
+        LOG.trace("Configured with time zone: {}", timezone);
+        this.rawMessageParser = rawMessageParser;
+    }
+
+    @Nullable
+    @Override
+    public Message decode(@Nonnull RawMessage rawMessage) {
+        String s = new String(rawMessage.getPayload(), StandardCharsets.UTF_8);
+        LOG.trace("Received raw message: {}", s);
+
+        PaloAltoMessageBase p = null;
+        try {
+            p = rawMessageParser.parse(s, timezone);
+        } catch (Exception e) {
+            LOG.warn("Cannot parse malformed Palo Alto 11x Message. Leaving message unparsed: {}", s);
+        }
+
+        String payload = s;
+        String source = getRawMessageSource(rawMessage);
+        DateTime timestamp = DateTime.now(DateTimeZone.UTC);
+        String panType = UNKNOWN;
+        if (p != null) {
+            payload = firstNonNull(p.payload(), payload);
+            source = firstNonNull(p.source(), source);
+            timestamp = firstNonNull(p.timestamp(), DateTime.now(DateTimeZone.UTC));
+            panType = firstNonNull(p.panType(), UNKNOWN);
+        }
```

Comment:
I think setting the default values two times here (once initially above, and once if the parsed value is null) is a bit confusing and is potentially error prone. What do you think about the following approach instead, which only overrides with the value if present, and leaves the default value in-place if not? 
```suggestion
if (p != null) {
            if (p.payload() != null) {
                payload = p.payload();
            }
            if (p.source() != null) {
                source = p.source();
            }
         ...
        }
```